File: internal/agent/actions/route_lookup.go
package actions

import (
	"context"
	"encoding/json"
	"slices"
	"strings"

	"github.com/Jkaotlic/wg-monitor/internal/agent/awgmgr"
	"github.com/Jkaotlic/wg-monitor/pkg/wire"
)

// Коды Notes ответа route_lookup. Слова к ним подбирает экран: агент говорит
// фактами, а не прозой.
const (
	lookupNoteHRNotRunning     = "hr_not_running"
	lookupNotePoliciesUnknown  = "policies_unknown"
	lookupNoteSingboxRouter    = "singbox_router"
	lookupNoteIPRulesUnchecked = "ip_rules_unchecked"
	lookupNoteRegexpUnchecked  = "regexp_unchecked"
	lookupNoteGeoExpandFailed  = "geo_expand_failed:" // + тег
)

// RouteLookup answers where one site goes according to the router's own rules
// and returns the answer JSON-encoded for wire.CommandResult.Output.
//
// Read-only: it reads the same inputs as RouteStatus and asks awg-manager to
// expand geosite tags, nothing else.
func RouteLookup(ctx context.Context, c *awgmgr.Client, domain string) (string, error) {
	in, err := fetchRouteInputs(ctx, c)
	if err != nil {
		return "", err
	}
	res := lookupRoute(domain, in, func(tag string) ([]string, error) {
		return c.GeoExpand(ctx, "geosite", tag)
	})
	b, err := json.Marshal(res)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// lookupRoute is the pure core of route_lookup: which enabled rules name the
// site and where each of them leads, by the very binding route_status credits
// the rule with (ruleEgress). No rule naming the site means the router's main
// exit carries it -- and that exit may well be a VPN tunnel, never assumed to
// be the provider.
//
// expand is called at most once per tag within one call.
func lookupRoute(domain string, in routeInputs, expand func(tag string) ([]string, error)) wire.RouteLookupResult {
	host := strings.TrimRight(strings.ToLower(strings.TrimSpace(domain)), ".")
	res := wire.RouteLookupResult{Domain: host, Matches: []wire.RouteLookupMatch{}}

	// sing-box раскладывает трафик по своим правилам, и ответ по правилам
	// роутера был бы уверенной неправдой.
	if in.settings.SingboxRouterActive() {
		res.Verdict = wire.LookupViaUnknown
		res.Notes = []string{lookupNoteSingboxRouter}
		return res
	}

	snap, env := routeSnapshotBase(in.hr, in.tunnels, in.routing, activeDefaultTunnelID(in.settings), in.policies, in.polIfaces, in.policiesUnknown)
	// hr == nil -- состояние движка не прочиталось. Объявить его правила
	// недействующими по такому поводу значило бы выдать догадку за факт.
	hrStopped := in.hr != nil && !in.hr.Running
	geo := &geoCache{expand: expand, cache: map[string]geoListEntry{}}
	var notes lookupNotes

	for _, r := range in.dns {
		if !r.Enabled {
			continue
		}
		if hrStopped && isHydraRouteBackend(r) {
			notes.add(lookupNoteHRNotRunning)
			continue
		}
		pattern, unchecked := ruleNamesHost(host, r, geo)
		if pattern == "" {
			// Непроверенное в силе, только пока правило сайт не назвало.
			notes.add(unchecked...)
			continue
		}
		_, tunnelID, known, _ := ruleEgress(r, env)
		via, id, name := env.egressVia(tunnelID, known)
		if via == wire.LookupViaUnknown && in.policiesUnknown {
			notes.add(lookupNotePoliciesUnknown)
		}
		res.Matches = append(res.Matches, wire.RouteLookupMatch{
			RuleName: firstNonEmptyRoute(r.Name, r.ID), Pattern: pattern,
			Via: via, TunnelID: id, TunnelName: name,
		})
	}
	// Статический маршрут -- правило по адресу сети: по имени сайта его не
	// проверить, а сайт может оказаться за ним.
	for _, s := range in.statics {
		if s.Enabled && len(s.Subnets) > 0 {
			notes.add(lookupNoteIPRulesUnchecked)
			break
		}
	}

	if len(res.Matches) == 0 {
		res.ByDefault = true
		res.Verdict, res.TunnelID, res.TunnelName = env.defaultVia(snap.DefaultEgress)
	} else {
		first := res.Matches[0]
		res.Verdict, res.TunnelID, res.TunnelName = first.Via, first.TunnelID, first.TunnelName
		for _, m := range res.Matches[1:] {
			if m.Via != first.Via || m.TunnelID != first.TunnelID {
				res.Verdict, res.TunnelID, res.TunnelName = wire.LookupMixed, "", ""
				break
			}
		}
	}
	if len(notes) > 0 {
		res.Notes = notes
	}
	return res
}

// egressVia turns a rule's egress (ruleEgress) into the answer's vocabulary.
// A known exit that is not a tunnel of ours is the provider; a tunnel id the
// snapshot does not show is unknown -- naming a line the screen cannot show
// would pass a guess off as a fact.
func (e *routeEgressEnv) egressVia(tunnelID string, known bool) (via, id, name string) {
	if !known {
		return wire.LookupViaUnknown, "", ""
	}
	if tunnelID == "" {
		return wire.LookupViaDirect, "", ""
	}
	t, ok := routeTunnelByID(e.tunnels, tunnelID)
	if !ok {
		return wire.LookupViaUnknown, "", ""
	}
	switch strings.ToLower(strings.TrimSpace(t.Type)) {
	case "managed", "system":
		return wire.LookupViaTunnel, t.ID, t.Name
	case "wan":
		return wire.LookupViaDirect, "", ""
	}
	return wire.LookupViaUnknown, "", ""
}

// defaultVia answers for a site no rule names: it goes wherever the router's
// main exit (RouteSnapshot.DefaultEgress) goes. "" is the router's own silence
// and stays unknown.
func (e *routeEgressEnv) defaultVia(defaultEgress string) (via, id, name string) {
	switch defaultEgress {
	case "":
		return wire.LookupViaUnknown, "", ""
	case wire.DefaultEgressDirect:
		return wire.LookupViaDirect, "", ""
	}
	return e.egressVia(defaultEgress, true)
}

// ruleNamesHost returns the first target of r that names host ("" when none
// does). unchecked lists the codes of targets a site name cannot be checked
// against; they matter only when the rule names nothing.
func ruleNamesHost(host string, r awgmgr.DNSRoute, geo *geoCache) (pattern string, unchecked []string) {
	targets := make([]string, 0, len(r.Domains)+len(r.ManualDomains)+len(r.Subnets))
	targets = append(targets, r.Domains...)
	targets = append(targets, r.ManualDomains...)
	targets = append(targets, r.Subnets...)
	note := func(code string) {
		if !slices.Contains(unchecked, code) {
			unchecked = append(unchecked, code)
		}
	}
	// linesName checks a list of v2ray-style lines; a regexp line is noted,
	// never guessed at.
	linesName := func(lines ...string) bool {
		for _, line := range lines {
			hit, isRegexp := geoLineNamesHost(host, line)
			if hit {
				return true
			}
			if isRegexp {
				note(lookupNoteRegexpUnchecked)
			}
		}
		return false
	}
	for _, raw := range targets {
		t := strings.TrimSpace(raw)
		lower := strings.ToLower(t)
		switch {
		case t == "":
		case strings.HasPrefix(lower, "geosite:"):
			tag := strings.TrimSpace(t[len("geosite:"):])
			lines, err := geo.lines(tag)
			if err != nil {
				// Не раскрытый список считается несовпавшим: сказать «идёт
				// через туннель» по списку, которого никто не видел, нельзя.
				note(lookupNoteGeoExpandFailed + tag)
				continue
			}
			if linesName(lines...) {
				return t, nil
			}
		case strings.HasPrefix(lower, "geoip:"):
			note(lookupNoteIPRulesUnchecked)
		default:
			switch target := normalizeRouteTarget(t); target.Type {
			case "cidr":
				note(lookupNoteIPRulesUnchecked)
			case "domain":
				if routeTargetMatchesHost(host, target.Value) {
					return t, nil
				}
			default:
				// ".claude.ai", "full:x" и прочее, что правило может нести
				// тем же синтаксисом, что и гео-список.
				if linesName(t) {
					return t, nil
				}
			}
		}
	}
	return "", unchecked
}

// geoLineNamesHost matches one line of a v2ray-style geosite list against
// host: ".x", "x" and "domain:x" name x and everything under it; "full:x" names
// x alone; "keyword:x" names any host containing x. "regexp:" lines are not
// evaluated (isRegexp=true) -- a pattern checked by a different engine than
// the router's would answer for the router without being it.
func geoLineNamesHost(host, line string) (hit, isRegexp bool) {
	l := strings.ToLower(strings.TrimSpace(line))
	switch {
	case l == "" || strings.HasPrefix(l, "#"):
		return false, false
	case strings.HasPrefix(l, "regexp:"):
		return false, true
	}
	// Атрибуты списка ("x @ads", "x:@ads") к имени не относятся.
	if i := strings.IndexAny(l, " \t@"); i >= 0 {
		l = strings.TrimRight(strings.TrimSpace(l[:i]), ":")
	}
	switch {
	case strings.HasPrefix(l, "full:"):
		return host == strings.TrimPrefix(l, "full:"), false
	case strings.HasPrefix(l, "keyword:"):
		kw := strings.TrimPrefix(l, "keyword:")
		return kw != "" && strings.Contains(host, kw), false
	}
	suffix := strings.Trim(strings.TrimPrefix(l, "domain:"), ".")
	return suffix != "" && (host == suffix || strings.HasSuffix(host, "."+suffix)), false
}

// geoCache expands each geosite tag at most once per lookup: one rule set
// routinely names the same tag from several rules, and every expansion is a
// round-trip to the router.
type geoCache struct {
	expand func(tag string) ([]string, error)
	cache  map[string]geoListEntry
}

type geoListEntry struct {
	lines []string
	err   error
}

func (g *geoCache) lines(tag string) ([]string, error) {
	if e, ok := g.cache[tag]; ok {
		return e.lines, e.err
	}
	lines, err := g.expand(tag)
	g.cache[tag] = geoListEntry{lines: lines, err: err}
	return lines, err
}

// lookupNotes keeps codes in first-seen order, each once.
type lookupNotes []string

func (n *lookupNotes) add(codes ...string) {
	for _, c := range codes {
		if !slices.Contains(*n, c) {
			*n = append(*n, c)
		}
	}
}

File: miniapp/src/routeLookup.js
// «Куда пойдёт сайт»: чистые функции раздела на экране маршрутов. Агент
// отвечает на route_lookup фактами и кодами (wire.RouteLookupResult), слова
// для владельца подбираются здесь -- о последствиях, без внутренней кухни.

const NOTE_TEXT = {
  hr_not_running: 'HydraRoute Neo — движок умной раздельной маршрутизации — сейчас не запущен, его правила не действуют',
  ip_rules_unchecked: 'Есть правила по адресу — по имени сайта их не проверить',
  regexp_unchecked: 'Часть правил записана шаблоном — их не проверить',
  policies_unknown: 'Роутер не отдал общие наборы правил',
  singbox_router: 'Трафиком управляет sing-box — он решает сам',
}
const GEO_EXPAND_FAILED = 'geo_expand_failed:'
const EXIT_UNRECOGNIZED = 'exit_unrecognized:'

// NOT_A_SITE -- одни и те же слова для имени, отбитого проверкой на экране,
// сервером (400 invalid_domain) и агентом.
export const NOT_A_SITE = 'Это не похоже на адрес сайта — нужно имя вроде claude.ai'

// normalizeSiteInput оставляет от того, что вставил человек, одно имя сайта:
// без схемы, пути, запроса, якоря, порта и хвостовой точки, в нижнем
// регистре. www. не трогается -- это другое имя, и правила бывают на нём.
export function normalizeSiteInput(raw) {
  let s = String(raw ?? '').trim().toLowerCase()
  s = s.replace(/^https?:\/\//, '')
  s = s.split(/[/?#]/, 1)[0]
  s = s.replace(/:\d*$/, '')
  return s.replace(/\.+$/, '').trim()
}

// looksLikeSite -- та же проверка, что у сервера (sanitizeWizardCommandArgs,
// route_lookup): не пусто, не длиннее 253, есть точка, нет пробелов, «/» и
// «:». Отбитое сервером не отправляется, а человек сразу слышит, что не так.
// Решает всё равно сервер: это подсказка, а не граница.
export function looksLikeSite(domain) {
  const d = String(domain ?? '')
  return d.length > 0 && d.length <= 253 && d.includes('.') && !/[\s/:]/.test(d)
}

function noteText(code) {
  if (typeof code !== 'string') return ''
  if (code.startsWith(GEO_EXPAND_FAILED)) {
    return `Роутер не раскрыл список «${code.slice(GEO_EXPAND_FAILED.length)}»`
  }
  if (code.startsWith(EXIT_UNRECOGNIZED)) {
    return `Сайт уйдёт через подключение «${code.slice(EXIT_UNRECOGNIZED.length)}» — роутер не сказал, VPN-туннель это или провайдер`
  }
  return NOTE_TEXT[code] ?? ''
}

function parseLookup(result) {
  if (typeof result === 'string') {
    try {
      return parseLookup(JSON.parse(result))
    } catch {
      return null
    }
  }
  return result && typeof result === 'object' ? result : null
}

function destination(via, tunnelName) {
  if (via === 'tunnel') return `через VPN-туннель «${tunnelName}»`
  if (via === 'direct') return 'напрямую через провайдера'
  return 'неизвестно куда'
}

// lookupAnswer turns a route_lookup result (object or the command's JSON
// output) into { title, lines, tone }. tone: ok -- ответ есть; warn -- ответ
// есть, но в нём есть подвох; unknown -- ответа нет.
export function lookupAnswer(result) {
  const r = parseLookup(result)
  if (!r) {
    return {
      title: 'Не удалось узнать, куда пойдёт сайт',
      lines: ['Роутер прислал ответ, который не удалось разобрать'],
      tone: 'unknown',
    }
  }
  const site = `«${r.domain || 'сайт'}»`
  const tunnel = r.tunnel_name || r.tunnel_id
  const matches = Array.isArray(r.matches) ? r.matches : []
  const codes = Array.isArray(r.notes) ? r.notes : []
  const notes = [...new Set(codes.map(noteText).filter(Boolean))]
  const rules = [...new Set(matches.map((m) => `правило «${m.rule_name}»`))]
  // Остановленный движок значит, что правила, на которые человек, возможно,
  // рассчитывает, не действуют: ответ верный, но повод насторожиться.
  const tone = codes.includes('hr_not_running') ? 'warn' : 'ok'

  switch (r.verdict) {
    case 'tunnel':
      return {
        title: r.by_default
          ? `Правил для ${site} нет — сайт пойдёт главным выходом роутера, через VPN-туннель «${tunnel}»`
          : `${site} пойдёт через VPN-туннель «${tunnel}»`,
        lines: [...rules, ...notes],
        tone,
      }
    case 'direct':
      return {
        title: r.by_default
          ? `Правил для ${site} нет — сайт пойдёт напрямую через провайдера`
          : `${site} пойдёт напрямую через провайдера`,
        lines: [...rules, ...notes],
        tone,
      }
    case 'mixed':
      return {
        title: `Несколько правил ведут ${site} в разные места`,
        lines: [
          ...matches.map((m) => `правило «${m.rule_name}» — ${destination(m.via, m.tunnel_name || m.tunnel_id)}`),
          ...notes,
        ],
        tone: 'warn',
      }
    default: {
      // Причина обязана прозвучать: «неизвестно» без «почему» человеку
      // ничего не даёт.
      const why = notes.length
        ? notes
        : [r.by_default ? 'Роутер не назвал свой главный выход' : 'Роутер не сказал, куда ведёт это правило']
      return { title: `Не удалось узнать, куда пойдёт ${site}`, lines: [...rules, ...why], tone: 'unknown' }
    }
  }
}

// lookupRefusal -- слова для отказа: агент и сервер отвечают по-английски и
// для людей не пишут. code -- код ошибки сервера (useCommand errorCode):
// текст ошибки у api.js кода не несёт, и по одному тексту отказ сервера в
// имени читался бы как «роутер не ответил».
export function lookupRefusal(output, code) {
  if (code === 'invalid_domain' || /invalid[ _]domain/i.test(String(output ?? ''))) return NOT_A_SITE
  return 'Роутер не ответил на вопрос — попробуйте ещё раз'
}
